docs/design/debug-api.md

```diff
@@ -0,0 +1,352 @@
+# Debug API design
+
+## Purpose
+
+The Debug API gives you access to several non-standard RPC methods, which will allow you to inspect, debug and set certain debugging flags during runtime. Also, it could help you replay all the transactions that have been executed prior and save considerable effort for your debugging process.
+
+## Goals
+
+1. **Debugging Capabilities**: Enable developers to retrieve detailed information about transactions, execution traces, and other debugging-related data.
+2. **Enhanced Insights**: Provide deeper insights into the behavior of smart contracts, transaction execution, and state changes on the Hedera network.
+3. **Ethereum Equivalent**: Makes the Hedera JSON-RPC Relay more equivalent to other providers used in the ethereum ecosystem, by providing the needed endpoints for easier debugging by developers.
+
+## Architecture
+
+New Debug API methods can be added and handled by adding debug service, which will be blueprinted by new `debug` interface added in the relay interface itself. Each method will be separate from the other and they won't overlap in terms of functionality.
+
+### Interface
+```javascript
+export interface Debug {
+  traceTransaction(transactionHash, { tracer, tracerConfig });
+
+  getModifiedAccountsByHash(startHash, endHash);
+
+  getModifiedAccountsByNumber(startNum, endNum);
+}
+```
+
+### Service class
+```javascript
+class DebugService implements Debug{
+  traceTransaction(transactionHash, { tracer, tracerConfig }) {
+    // Implementation of traceTranscation depending on passed params.
+  }
+
+  getModifiedAccountsByHash(startHash, endHash){ 
+    // Implementation of getModifiedAccountsByHash depending on passed params.
+  }
+
+  getModifiedAccountsByNumber(startNum, endNum){
+    // Implementation of getModifiedAccountsByNumber depending on passed params.
+  }
+}
+```
+
+### Method description
+
+`debug_traceTransaction` - Attempts to run the transaction in the exact same manner as it was executed on the network.
```

Comment:
Please add a high level description of the logic to enable this in terms of other calls. That is what mirror node calls will the logic be making and what data formatting might be necessary

Reply:
Hi Nana, I also explored the possibility of implementing debug_traceTransaction. Unfortunately, it appears that the actions endpoint in the mirrornode ->  /api/v1/contracts/results/{transactionIdOrHash}/actions does not support all opcodes and the information it returns is not enough to return an adequate trace for a given transaction. We would need change in the mirror node in order to make this work

Reply:
If we cannot support `debug_traceTransaction` that means we cannot support `trace_filter` either?

Reply:
Yes, this is in line with expectations.
Initially, we will only support the callTrace form of `debug_traceTransaction` which should only need the side car actions.
Once the opcode flow of [HIP 801](https://github.com/hashgraph/hedera-improvement-proposal/pull/801) is implemented we'll simply make the call to `/api/v1/contracts/results/{transactionIdOrHash}/opcodes`

I just want the design doc to call out that when
- `tracer = callTracer` then we call `/api/v1/contracts/results/{transactionIdOrHash}/actions` and format the response
- `tracer = opcodeLogger` then we call `/api/v1/contracts/results/{transactionIdOrHash}/opcodes` and format the response

Then it should briefly describe any intricate response formatting needed. If it's just basic number to hex etc there's no need to call that part out in detail